torchprime/sharding/tests/test_shard_model.py

```diff
@@ -0,0 +1,237 @@
+import torch
+import torch.nn as nn
+
+from torchprime.sharding.shard_model import (
+  ShardedModule,
+  shard_model,
+  shard_model_from_config,
+)
+
+
+class SimpleLinear(nn.Module):
+  def __init__(self):
+    super().__init__()
+    self.fc1 = nn.Linear(128, 64)
+    self.relu = nn.ReLU()
+    self.fc2 = nn.Linear(64, 128)
+
+  def forward(self, x):
+    y = self.relu(self.fc1(x))
+    z = self.fc2(y)
+    return z
+
+
+class MockShardedTensor(torch.Tensor):
+  """
+  This class simulates a sharded tensor.
+  """
+
+  def __init__(self, orig):
+    super().__init__()
+    self.orig = orig
+
+
+class MockShardedModule(nn.Module):
+  """
+  This class simulates an activation (output) sharded module.
+  """
+
+  def __init__(self, orig):
+    super().__init__()
+    self.orig = orig
+
+  def forward(self, x):
+    return self.orig(x)
+
+
+def test_traverse_weights():
+  model = SimpleLinear()
+  visited = set()
+
+  def shard_weight(weight, name):
+    visited.add(name)
+    return MockShardedTensor(weight)
+
+  model = shard_model(model, shard_weight, lambda x, _: x)
+  assert visited == {"fc1.weight", "fc1.bias", "fc2.weight", "fc2.bias"}
+
+  # Check that all weights are sharded.
+  sharded = set()
+  for name, param in model.named_parameters():
+    assert isinstance(param.data, MockShardedTensor)
+    sharded.add(name)
+  assert sharded == visited
+
+
+def test_traverse_modules():
+  model = SimpleLinear()
+  visited = set()
+
+  def shard_activation(module, name):
+    visited.add(name)
+    return MockShardedModule(module)
+
+  model = shard_model(model, lambda x, _: x, shard_activation)
+  assert visited == {"", "fc1", "relu", "fc2"}
+
+  # Check that all modules are sharded.
+  sharded = set()
+  for name, mod in model.named_modules():
+    if isinstance(mod, MockShardedModule):
+      sharded.add(name)
+  assert len(sharded) == len(visited)
```

Comment:
Wasn't expecting an empty name in the list while going over named_modules but good to know.

Reply:
Commented